bee-tangle/src/ms_tangle.rs

```diff
@@ -0,0 +1,1145 @@
+    entry_point_index: AtomicU32,
+    tip_pool: Mutex<UrtsTipPool>,
+}
+
+impl<B> Deref for MsTangle<B> {
+    type Target = Tangle<MessageMetadata, StorageHooks<B>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<B: StorageBackend> MsTangle<B> {
+    /// Create a new `MsTangle` instance with the given configuration and storage handle.
+    pub fn new(config: TangleConfig, storage: ResourceHandle<B>) -> Self {
+        Self {
+            config,
+            inner: Tangle::new(StorageHooks { storage }),
+            milestones: Default::default(),
+            solid_entry_points: Default::default(),
+            latest_milestone_index: Default::default(),
+            solid_milestone_index: Default::default(),
+            confirmed_milestone_index: Default::default(),
+            snapshot_index: Default::default(),
+            pruning_index: Default::default(),
+            entry_point_index: Default::default(),
+            tip_pool: Mutex::new(UrtsTipPool::default()),
+        }
+    }
+
+    /// Shut down the tangle, terminating any and all worker tasks.
+    pub async fn shutdown(self) {
+        // TODO: Write back changes by calling self.inner.shutdown().await
+    }
+
+    /// Get the configuration of this tangle.
+    pub fn config(&self) -> &TangleConfig {
+        &self.config
+    }
+
+    /// Insert a message into the tangle.
+    pub async fn insert(&self, message: Message, hash: MessageId, metadata: MessageMetadata) -> Option<MessageRef> {
+        // TODO this has been temporarily moved to the processor.
+        // Reason is that since the tangle is not a worker, it can't have access to the propagator tx.
+        // When the tangle is made a worker, this should be put back on.
+        // if opt.is_some() {
+        //     if let Err(e) = Protocol::get()
+        //         .propagator_worker
+        //         .unbounded_send(PropagatorWorkerEvent(hash))
+        //     {
+        //         error!("Failed to send hash to propagator: {:?}.", e);
+        //     }
+        // }
+        //
+        // opt
+        self.inner.insert(hash, message, metadata).await
+    }
+
+    /// Add a milestone to the tangle.
+    pub async fn add_milestone(&self, idx: MilestoneIndex, milestone: Milestone) {
+        // TODO: only insert if vacant
+        self.inner
+            .update_metadata(&milestone.message_id(), |metadata| {
+                metadata.flags_mut().set_milestone(true);
+                metadata.set_milestone_index(idx);
+                metadata.set_omrsi(IndexId::new(idx, *milestone.message_id()));
+                metadata.set_ymrsi(IndexId::new(idx, *milestone.message_id()));
+            })
+            .await;
+        self.inner
+            .hooks()
+            .insert_milestone(idx, &milestone)
+            .await
+            .unwrap_or_else(|e| info!("Failed to insert message {:?}", e));
+        // TODO can there be a race condition between 2 ops ?
```

Comment:
I'd add a log entry before and after acquiring the lock just to be sure.

Reply:
Maybe we could have a wrapper for this. @Adam-Gleave might have some opinions about this.

Reply:
If we're not sure something bad could happen here, probably best if we do something along those lines in case we happen to run into it. Would save a lot of time debugging. Something like the wrapper we were using for `wallet.rs` may be a little heavyweight, though.

Reply:
This is actually no longer relevant because the implementation has changed such that there are no longer lock dependencies (i.e: there's no chance of a deadlock).